messages: Copy a markdown text from multiple messages.

afreydev opened the pull request:
This is a new feature for copying to clipboard a markdown
text from a set of selected messages. This is related with #13936 

The user can use:
* Ctrl + click, for selecting (or deselecting) an individual
message.
* Shift + click, for selecting a range of messages from
the current selected message (blue board) to the message
that was clicked.

This change added:
* Mouse events for select messages (click_handlers.js)
* JS module for managing the actions (message_copy.js).
* A new option in the message popover menu.
This include a template and its styles.
* A new endpoint for getting the markdown text.
It is different from /messages (GET), because this is for a
list of messages and it includes additional info like:
message datetime, a sender silent mention
and the message.
* A management command with the same functionality.
* Test for backend actions.
* Test for the new frontend module.
* OpenAPI documentation for the new endpoint.

This is the new feature in action. Ctrl + click for selecting a message and Shift + click for selecting 
the range between the current selected message (blue border) and the clicked message.

![copy-paste-f](https://user-images.githubusercontent.com/43050896/75474678-6805c900-5965-11ea-8a34-ed19348b4495.gif)

For the backend, I have tested:

* Sending a list of message_id and getting the raw text.
* Sendind invalid ids.
* Sending an empty list.
* Sending a list with unauthorized/private messages.
* Sending a list to the endpoint /messages/markdown

For the frontend, I have used zjquery for testing the JS module message_copy,js.
* Select and deselect a message.
* Select multiple messages.
* Clean UI.
* Copy messages.

afreydev commented:
This is the zulip stream for this feature: https://chat.zulip.org/#narrow/stream/137-feedback/topic/copy.20paste

timabbott commented on zerver/lib/actions.py:
```diff
@@ -5912,3 +5912,17 @@ def do_delete_realm_export(user_profile: UserProfile, export: RealmAuditLog) ->
+    buffer = []
+    for message_id in messages:
+        try:
+            (message, user_message) = access_message(user_profile, message_id)
```
This should use `bulk_access_messages`.  While that's currently no better than what you have, it'll mean improving that function fixes this code path as well.

afreydev replied:
Tim, this method has a param a sequence of Message:

bulk_access_messages(user_profile: UserProfile, messages: Sequence[Message]) -> List[Message]:

We have an int List.

It would be an additional process (create messages with the id or find messages by id).

pirate replied:
We could pass a generator to that existing function:

```python3
bulk_access_messages(user_profile, (access_message(user_profile, message_id)[0] for message_id in messages))
```

Or we could update the existing function to also take Message ID sequences:

```python3
def bulk_access_messages(user_profile: UserProfile, messages: Union[Sequence[Message], Sequence[int]]) -> List[Message]:
    filtered_messages = []

    for message_or_id in messages:
        if isinstance(message_or_id, Message):
            message = message_or_id
            user_message = get_usermessage_by_message_id(user_profile, message.id)
        elif isinstance(message_or_id, int):
            message, user_message = access_message(user_profile, message_id)
        else:
            raise TypeError('messages argument must be a Sequence[Message] or Sequence[int] of Message IDs')
            
        if has_message_access(user_profile, message, user_message):
            filtered_messages.append(message)
    return filtered_messages
```

I'm kinda meh about both approaches over the existing one, idk if there's a better way.

timabbott replied:
`bulk_access_messages` has I think only one real caller and only needs the IDs, not the objects; it's already fetching the messages as `user_message.message` anyway in `get_usermessage_by_message_id`.  Just migrate its API as a prep commit.

(The main thing we use for this kind of application is `bulk_access_messages_expect_usermessage`, but that's not appropriate for this use case)

timabbott commented on zproject/urls.py:
```diff
@@ -202,6 +202,8 @@
     url(r'^messages/matches_narrow$', rest_dispatch,
         {'GET': 'zerver.views.messages.messages_in_narrow_backend'}),
+    url(r'^messages/markdown$', rest_dispatch,
+        {'GET': 'zerver.views.messages.json_fetch_raw_list_messages'}),
```
Why do we need a new API endpoint?  We should be able to do everything we want with using `GET /messages` to fetch the messages (passing `apply_markdown=False` to get the raw content), and then having the client do the little bit of formatting of sender names, date, etc.

afreydev replied:
Hi Tim, the main reason is the performance. For example, if you need the data of a message, it wouldn't important. But, if you have a range of 10, 20 or >30  messages, it would be the same number of requests to the server. In the frontend, it would be slow too.

afreydev replied:
@timabbott, if you are telling me about:
^messages$ -> 'GET': 'zerver.views.messages.get_messages_backend'
It works with a homogeneus range, but we can have a list with individual messages mixed with little ranges too.

pirate replied:
Should we add a new option to the existing endpoint to allow it to accept individual ranges & message ids instead of adding a new endpoint?

timabbott replied:
I'm a little puzzled here -- Is that idea that one will be able to manually select/deselect individual messages in the UI?  If it's a range of things within a view that the Zulip webapp can display, there's a `get_messages` query that will return that range of messages.

I would potentially be open to add a "Get messages by ID" option to `get_messages` where you just provide a list of message IDs and no anchor or anything, but I'm not sure it's relevant to what you're trying to do here.

afreydev replied:
The feature requires something like the second option ("Get messages by ID"), because we are sending to the endpoint an message ID array.
If this is possible:
1. The view is "get_messages_backend", isn't it?

`def get_messages_backend(request: HttpRequest, user_profile: UserProfile,
                         anchor_val: Optional[str]=REQ(
                             'anchor', str_validator=check_string, default=None),
                         num_before: int=REQ(converter=to_non_negative_int),
                         num_after: int=REQ(converter=to_non_negative_int),
                         narrow: OptionalNarrowListT=REQ('narrow', converter=narrow_parameter, default=None),
                         use_first_unread_anchor_val: bool=REQ('use_first_unread_anchor',
                                                               validator=check_bool, default=False),
                         client_gravatar: bool=REQ(validator=check_bool, default=False),
                         apply_markdown: bool=REQ(validator=check_bool, default=True)) -> HttpResponse`

2. num_before and num_after would have a default value?
3. should we add a boolean param "get_by_ids" and a new param "message_ids"?

timabbott replied:
@afreydev I'm still not clear on why the frontend here needs "Get messages by ID" -- can you answer that question clearly?

afreydev replied:
For this feature, the user can select "continuous or discontinuous" group of messages. If you check the gif, sometimes you can select 1 or 2 or 3 consecutive messages but in other cases select multiple not consecutive messages. 
I have updated the version, in this I have deleted the new endpoint but I have added the param message_id_list to get_messages_backend. Couldn't it work like that?

afreydev replied:
![copy-paste4](https://user-images.githubusercontent.com/43050896/75832613-9bd95800-5d84-11ea-9001-b06845471ed8.gif)

afreydev commented:
Hello @timabbott , is it possible rerun bionic-backend-python3.6? The fail is caused by something about a github user.

timabbott commented:
@afreydev yeah, one can rerun; though you could also just rebase and push again.

afreydev committed:
messages: Copy a markdown text from multiple messages.

This is a new feature for copying to clipboard a markdown
text from a set of selected messages.

The user can use:
* Ctrl + click, for selecting (or deselecting) an individual
message.
* Shift + click, for selecting a range of messages from
the current selected message (blue board) to the message
that was clicked.

This change added:
* Mouse events for select messages (click_handlers.js)
* JS module for managing the actions (message_copy.js).
* A new option in the message popover menu.
This include a template and its styles.
* A new parameter called message_id_list in
get_messages_backend view.
It is to get messages by id.
* Test for backend actions.
* Test for the new frontend module.
* OpenAPI documentation for the new parameter.

alya commented:
@apoorvapendse Since #13936 is resolved, should this PR be closed now?

apoorvapendse commented:
Hey @alya, I think this PR can be closed, given that it is quite old.
I'm not sure why the original issue was closed on following the issue thread.

> For this feature, the user can select "continuous or discontinuous" group of messages.

The feature proposed here is not completely implemented.

We have a subset of that implemented, i.e. being able to select continuous messages and copy the markdown to clipboard.
I think we are improving that aspect to extend continuous selection to partially select messages or do a better overlap of selected content vs copied content through #35100.

But the Shift + Up/Down Arrow shortcut doesn't do a continuous selection, which seems to be a part of feature set and the linked CZO discussion.
And discontinuous selection doesn't exist.
Maybe we can file a new issue to track outstanding features, if we still want them.

apoorvapendse commented:
[#design > select messages to move/delete](https://chat.zulip.org/#narrow/channel/101-design/topic/select.20messages.20to.20move.2Fdelete)
Might also be relevant here.

apoorvapendse reviewed (commented):
Hey @alya, I think this PR can be closed, given that it is quite old.
I'm not sure why the original issue was closed on following the issue thread.

> For this feature, the user can select "continuous or discontinuous" group of messages.

The feature proposed here is not completely implemented.

We have a subset of that implemented, i.e. being able to select continuous messages and copy the markdown to clipboard.
I think we are improving that aspect to extend continuous selection to partially select messages or do a better overlap of selected content vs copied content through #35100.

But the Shift + Up/Down Arrow shortcut doesn't do a continuous selection, which seems to be a part of feature set and the linked [CZO discussion](https://chat.zulip.org/#narrow/stream/137-feedback/topic/copy.20paste).
And discontinuous selection doesn't exist.
Maybe we can file a new issue to track outstanding features, if we still want them.

[#design > select messages to move/delete](https://chat.zulip.org/#narrow/channel/101-design/topic/select.20messages.20to.20move.2Fdelete)
Might also be relevant here.

alya commented:
Thanks for digging up that history! I don't think discontinuous selection for copying is a priority; you *can* quote a bunch of messages into the compose box (one by one). We can think about it after we figure out the checkbox UI for moving and deleting messages.

alya commented:
Closing this PR as #13936 is closed and much has changed since it was created. Thanks for the work here, @afreydev !
